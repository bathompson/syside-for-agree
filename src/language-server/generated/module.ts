/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

/******************************************************************************
 * This file was generated by langium-cli 0.5.0.
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/

import { LangiumGeneratedServices, LangiumGeneratedSharedServices, LangiumSharedServices, LangiumServices, LanguageMetaData, Module, IParserConfig } from 'langium';
import { SysMlAstReflection } from './ast';
import { KerMLGrammar, SysMLGrammar } from './grammar';

export const KerMLLanguageMetaData: LanguageMetaData = {
    languageId: 'kerml',
    fileExtensions: ['.kerml'],
    caseInsensitive: false
};

export const SysMLLanguageMetaData: LanguageMetaData = {
    languageId: 'sysml',
    fileExtensions: ['.sysml'],
    caseInsensitive: false
};

export const parserConfig: IParserConfig = {
    recoveryEnabled: true,
    nodeLocationTracking: 'none',
};

export const SysMlGeneratedSharedModule: Module<LangiumSharedServices, LangiumGeneratedSharedServices> = {
    AstReflection: () => new SysMlAstReflection()
};

export const KerMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => KerMLGrammar(),
    LanguageMetaData: () => KerMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};

export const SysMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => SysMLGrammar(),
    LanguageMetaData: () => SysMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};
