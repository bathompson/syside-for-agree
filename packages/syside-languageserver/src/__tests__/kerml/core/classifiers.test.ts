/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

import { Classifier } from "../../../generated/ast";
import { anything, defaultLinkingErrorTo, parsedNode, qualifiedTarget } from "../../../testing";

describe.each(["specializes", "conjugates"])(
    "classifiers can only specialize other classifiers with '%s'",
    (token: string) => {
        test("specializing non-classifiers issues a diagnostic", async () => {
            return expect(
                `namespace A;
        classifier Child ${token} A;`
            ).toParseKerML(
                {},
                {
                    diagnostics: [defaultLinkingErrorTo("A"), ...anything(1)],
                    buildOptions: { validationChecks: "all", standardLibrary: "none" },
                }
            );
        });

        test("specializing classifiers is successful", async () => {
            return expect(
                parsedNode(
                    `classifier A;
        classifier Child ${token} A;`,
                    { node: Classifier, index: 1, build: true }
                )
            ).resolves.toMatchObject({
                heritage: [{ targetRef: qualifiedTarget("A") }],
            });
        });
    }
);

test("specializations can be parsed", async () => {
    return expect(`
    classifier A;
    classifier B;
    specialization Super subclassifier A specializes B;
    specialization subclassifier B :> A {
        doc /* unnamed */
    }`).toParseKerML();
});

test("classifiers can specialize multiple other classifiers", async () => {
    return expect(
        parsedNode(
            `
    classifier A;
    classifier B;
    classifier C specializes A, B;`,
            { node: Classifier, index: 2 }
        )
    ).resolves.toMatchObject({ heritage: anything(2) });
});
