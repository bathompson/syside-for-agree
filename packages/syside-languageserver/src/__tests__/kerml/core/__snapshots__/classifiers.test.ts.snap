// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classifiers can specialize multiple other classifiers: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Classifier",
          "children": [],
          "declaredName": "A",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Classifier",
          "children": [],
          "declaredName": "B",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Classifier",
          "children": [],
          "declaredName": "C",
          "heritage": [
            {
              "$type": "Subclassification",
              "elements": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Classifier",
                      "children": [],
                      "declaredName": "A",
                      "heritage": [],
                      "isSufficient": false,
                      "prefixes": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
            {
              "$type": "Subclassification",
              "elements": [],
              "reference": {
                "$type": "ClassifierReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Classifier",
                      "children": [],
                      "declaredName": "B",
                      "heritage": [],
                      "isSufficient": false,
                      "prefixes": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;

exports[`specializations can be parsed: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "children": [
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Classifier",
          "children": [],
          "declaredName": "A",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Classifier",
          "children": [],
          "declaredName": "B",
          "heritage": [],
          "isSufficient": false,
          "prefixes": [],
          "typeRelationships": [],
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Subclassification",
          "declaredName": "Super",
          "elements": [],
          "reference": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "children": [],
                  "declaredName": "B",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "children": [],
                  "declaredName": "A",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
      },
      {
        "$type": "OwningMembership",
        "element": {
          "$type": "Subclassification",
          "elements": [
            {
              "$type": "Annotation",
              "element": {
                "$type": "Documentation",
                "about": [],
                "body": "/* unnamed */",
              },
              "elements": [],
            },
          ],
          "reference": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "children": [],
                  "declaredName": "A",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
          "source": {
            "$type": "ClassifierReference",
            "parts": [
              {
                "error": undefined,
                "ref": {
                  "$type": "Classifier",
                  "children": [],
                  "declaredName": "B",
                  "heritage": [],
                  "isSufficient": false,
                  "prefixes": [],
                  "typeRelationships": [],
                },
              },
            ],
          },
        },
        "elements": [],
      },
    ],
    "prefixes": [],
  },
}
`;
